lock-api/src/main/java/com/palantir/lock/client/LockTokenShare.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import com.google.common.base.Preconditions;
+import com.palantir.lock.v2.LockToken;
+
+final class LockTokenShare implements LockToken {
+    private final UUID requestId;
+    private final LockToken sharedLockToken;
+    private final ReferenceCounter referenceCounter;
+
+    private boolean unlocked;
+
+    private LockTokenShare(ReferenceCounter referenceCounter, LockToken token) {
+        this.referenceCounter = referenceCounter;
+        this.requestId = UUID.randomUUID();
+        this.sharedLockToken = token;
+        this.unlocked = false;
+    }
+
+    /**
+     * Returns a list of {@link LockTokenShare}, that are referencing to same LockToken.
+     *
+     * Share should be called only once on same {@link LockTokenShare} instance, guarantees do not hold between two
+     * different lists of {@link LockTokenShare} generated by successive share calls on same {@link LockToken}.
+     */
+    public static List<LockToken> share(LockToken token, int referenceCount) {
```

Comment:
all the 'public' declarations in this class are redundant because the class is package private

Reply:
will fix that in a separate pr